### PR TITLE
Debezium upgrade to 1.9.3.Final [HZ-1147]

### PR DESCRIPTION
**Pull request by TomaszGaweda**
This PR upgrades Debezium to 1.9.3.Final.

Breaking changes (list specific methods/types/messages):
* serialized form - use readString instead of readUTF, etc.
* snapshot format - use readString instead of readUTF, etc.

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [x] Send backports/forwardports if fix needs to be applied to past/future releases
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

**Comment by TomaszGaweda** [comment]
run-lab-run

**Comment by TomaszGaweda** [comment]
run-cdc-debezium-tests

**Comment by TomaszGaweda** [comment]
run-cdc-debezium-tests

**Review by frant-hartm — CHANGES_REQUESTED** [review]
Thanks, some nice cleanups there. I thought the upgrade will be more complicated.

I am not sure about repeating the `@Nonnull` annotations from the interface, I know intellij highlights it if we don't do that, but we don't do it across the code base.

File: extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/CdcSerializerHooks.java [hunk 1]
```diff
@@ -51,19 +52,21 @@ public int getTypeId() {
...
-                    out.writeUTF(record.getKeyJson());
-                    out.writeUTF(record.getValueJson());
+                    out.writeString(record.getKeyJson());
+                    out.writeString(record.getValueJson());
                 }
 
+                @Nonnull
                 @Override
-                public ChangeRecordImpl read(ObjectDataInput in) throws IOException {
+                @SuppressWarnings("ConstantConditions")
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/CdcSerializerHooks.java:64]: It's IntelliJ. Most of us use IntelliJ, I think it's a good idea to make code less "yellow" to easier see actual problems ;) However, I do not insist if it's not something we want to make

File: extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java [hunk 2]
```diff
@@ -48,33 +48,32 @@
...
 import static org.testcontainers.containers.PostgreSQLContainer.POSTGRESQL_PORT;
 
+@SuppressWarnings("SqlNoDataSourceInspection")
 @Category({NightlyTest.class})
 @RunWith(HazelcastSerialClassRunner.class)
 public class DebeziumCdcIntegrationTest extends AbstractCdcIntegrationTest {
 
     private static final DockerImageName MYSQL_IMAGE =
-            DockerImageName.parse("debezium/example-mysql:1.3").asCompatibleSubstituteFor("mysql");
+            DockerImageName.parse("debezium/example-mysql:1.9.1.Final").asCompatibleSubstituteFor("mysql");
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java:57]: The images did not change in any meaningful way. Keeping 1.9.1 makes us sure we are ready for 1.9 (once they fix the Java 8 support). Can change, but IMHO it's just "for the future" to say "we are ready"

File: extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java [hunk 3]
```diff
@@ -48,33 +48,32 @@
...
+        try (MySQLContainer<?> container = mySqlContainer()) {
             container.start();
 
             // given
             List<String> expectedRecords = Arrays.asList(
-                    "1001/0:INSERT:Customer {id=1001, firstName=Sally, lastName=Thomas, email=sally.thomas@acme.com}",
-                    "1002/0:INSERT:Customer {id=1002, firstName=George, lastName=Bailey, email=gbailey@foobar.com}",
-                    "1003/0:INSERT:Customer {id=1003, firstName=Edward, lastName=Walker, email=ed@walker.com}",
-                    "1004/0:INSERT:Customer {id=1004, firstName=Anne, lastName=Kretchmar, email=annek@noanswer.org}",
+                    "1001/0:SYNC:Customer {id=1001, firstName=Sally, lastName=Thomas, email=sally.thomas@acme.com}",
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java:73]: Previously the MySQL connector incorrectly was reporting initially loaded entities as just inserts; now it - correctly - says it's from initial load. The first update you see below is the first query we are doing after initial load.

File: extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java [hunk 4]
```diff
@@ -48,33 +48,32 @@
...
+                    "1002/0:SYNC:Customer {id=1002, firstName=George, lastName=Bailey, email=gbailey@foobar.com}",
+                    "1003/0:SYNC:Customer {id=1003, firstName=Edward, lastName=Walker, email=ed@walker.com}",
+                    "1004/0:SYNC:Customer {id=1004, firstName=Anne, lastName=Kretchmar, email=annek@noanswer.org}",
                     "1004/1:UPDATE:Customer {id=1004, firstName=Anne Marie, lastName=Kretchmar, email=annek@noanswer.org}",
                     "1005/0:INSERT:Customer {id=1005, firstName=Jason, lastName=Bourne, email=jason@bourne.org}",
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java:78]: It's after initial load was finished (it's not preloaded in the image, but inserted manually in the test)

File: extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/CdcSerializerHooks.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/CdcSerializerHooks.java]: I've replaced it with Objects.requireNonNull -> it still fixes highlighting (no false warnings) and adds additional debugging info, which Supression was not giving us.

**Comment by TomaszGaweda** [comment]
run-cdc-debezium-tests

**Comment by TomaszGaweda** [comment]
run-cdc-debezium-tests

**Comment by frant-hartm** [comment]
What is the issue with the newer postgres image?

Maybe we should bite the bullet and migrate to the official postgres image and do any required setup. I had issue with the debezium example image in some test, because it tries to recreate some data when you stop-start it in a middle of a test.

**Comment by TomaszGaweda** [comment]
Jenkins is reporting:
```
14:54:23,766 ERROR |customers| - [Final]] Time-limited test - Log output from the failed container:
ls: cannot access '/docker-entrypoint-initdb.d/': Operation not permitted
```

It requires [Docker upgrade](https://stackoverflow.com/questions/69706677/cannot-start-postgresql-docker-container-docker-entrypoint-initdb-d-oper). Today I asked on Slack again if upgrade is possible in near future

File: extensions/cdc-debezium/pom.xml [hunk 5]
```diff
@@ -33,7 +33,7 @@
     </parent>
 
     <properties>
-        <debezium.version>1.2.5.Final</debezium.version>
```
Comment (frant-hartm) [inline extensions/cdc-debezium/pom.xml:36]: Let's move this property to root pom.xml and have it only there.

File: extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcListenBeforeExistIntegrationTest.java [hunk 6]
```diff
@@ -54,6 +54,7 @@ public void listenBeforeDatabaseExists() throws Exception {
 
         StreamSource<ChangeRecord> source = sourceBuilder("cdcMysql")
                 .setDatabaseWhitelist(DATABASE)
+                .setCustomProperty("snapshot.mode", "never")
```
Comment (frant-hartm) [inline extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcListenBeforeExistIntegrationTest.java:57]: I have missed this one, why is this change needed?

**Review by frant-hartm — CHANGES_REQUESTED** [review]
Two more minor things, then it's good from my side.

File: extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcListenBeforeExistIntegrationTest.java (hunk 6 above)
Comment (TomaszGaweda) [reply extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/MySqlCdcListenBeforeExistIntegrationTest.java:57]: It's because we are inserting the rows just after connector started, we don't really know if the SYNC phase was done - and because of that, the result once will be INSERT, once SYNC. 

If we disable the initial snapshot, we will always have this record as INSERT, otherwise the snapshot process may read it before core binlog reader and report this as SYNC event.

Adding snapshot.mode = NEVER means that the snapshot process won't drop INSERT event in favor of SYNC, so the rest is predictable

**Comment by TomaszGaweda** [comment]
@frant-hartm Done :).

**Review by frant-hartm — APPROVED** [review]

**Comment by TomaszGaweda** [comment]
Thank you for the CR @frant-hartm and @ldziedziul :)
